lighthouse-core/scripts/pptr-run-devtools.js

```diff
@@ -0,0 +1,180 @@
+
+const argv = yargs(process.argv.slice(2))
+  .usage('$0 [url]')
+  .help('h').alias('h', 'help')
+  .option('_', {type: 'string'})
+  .option('o', {
+    type: 'string',
+    default: 'latest-run/devtools-lhrs',
+  }).alias('o', 'output-dir')
+  .option('d', {
+    type: 'string',
+  }).alias('d', 'custom-devtools-frontend')
+  .argv;
+
+/**
+ * https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/test_runner/TestRunner.js;l=170;drc=f59e6de269f4f50bca824f8ca678d5906c7d3dc8
+ * @param {Record<string, function>} receiver
+ * @param {string} methodName
+ * @param {function} override
+ */
+function addSniffer(receiver, methodName, override) {
+  const original = receiver[methodName];
+  if (typeof original !== 'function') {
+    throw new Error('Cannot find method to override: ' + methodName);
+  }
+
+  /**
+   * @param  {...any} args
+   */
+  receiver[methodName] = function(...args) {
+    let result;
+    try {
+      result = original.apply(this, args);
+    } finally {
+      receiver[methodName] = original;
+    }
+    // In case of exception the override won't be called.
+    try {
+      Array.prototype.push.call(args, result);
+      override.apply(this, args);
+    } catch (e) {
+      throw new Error('Exception in overriden method \'' + methodName + '\': ' + e);
+    }
+    return result;
+  };
+}
+
+const sniffLhr = `
+new Promise(resolve => {
+  (${addSniffer.toString()})(
+    UI.panels.lighthouse.__proto__,
+    '_buildReportUI',
+    (lhr, artifacts) => resolve(lhr)
+  );
+});
+`;
+
+const startLighthouse = `
+(async () => {
+  await UI.ViewManager.instance().showView('lighthouse');
+  const button = UI.panels.lighthouse.contentElement.querySelector('button');
+  if (button.disabled) throw new Error('Start button disabled');
+  button.click();
+})()
+`;
+
+/**
+ * @param {import('puppeteer').Browser} browser
+ * @param {string} url
+ * @return {Promise<string>}
+ */
+async function testPage(browser, url) {
+  const page = await browser.newPage();
+
+  const targets = await browser.targets();
+  const inspectorTarget = targets.filter(t => t.url().includes('devtools'))[1];
+  if (!inspectorTarget) throw new Error('No inspector found.');
+
+  const session = await inspectorTarget.createCDPSession();
+  await session.send('Runtime.enable');
+
+  // Navigate to page and wait for initial HTML to be parsed before trying to start LH.
+  await new Promise(async resolve => {
+    const pageSession = await page.target().createCDPSession();
+    await pageSession.send('Page.enable');
+    pageSession.once('Page.domContentEventFired', resolve);
+    page.goto(url).catch(err => err);
+  });
+
+  /** @type {RuntimeEvaluateResponse|undefined} */
+  let startLHResponse;
+  while (!startLHResponse || startLHResponse.exceptionDetails) {
+    startLHResponse = await session.send('Runtime.evaluate', {
+      expression: startLighthouse,
+      awaitPromise: true,
+    }).catch(err => err);
```

Comment:
does this `err` also have `exceptionDetails` on it? it's a little unclear if this will pass our `while` condition and if it's intentional that it should.

I was assuming the exceptionDetails was the [return object version](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#method-evaluate), but now I'm not sure :)

Reply:
Not sure, but it should force another execution of the loop. I think we can resolve this with:
```suggestion
    }).catch(err => ({exceptionDetails: err}));
```

Reply:
sg!